### PR TITLE
patch currently install CRDs

### PR DESCRIPTION
**Pull request by TheoBrigitte**
Towards: https://github.com/giantswarm/giantswarm/issues/20256

### CODE REVIEW

File: helm/prometheus-operator-app/templates/_helpers.tpl [hunk 1]
```diff
@@ -165,7 +165,21 @@ Use the prometheus-node-exporter namespace override for multi-namespace deployme
   {{- end -}}
 {{- end -}}
 
+{{- define "kube-prometheus-stack.crdInstall" -}}
+{{- printf "%s-%s" ( include "kube-prometheus-stack.name" . ) "crd-install" | replace "+" "_" | trimSuffix "-" -}}
+{{- end -}}
+
+{{- define "kube-prometheus-stack.CRDInstallAnnotations" -}}
+"helm.sh/hook": "pre-upgrade"
```
Comment (TheoBrigitte) [inline helm/prometheus-operator-app/templates/_helpers.tpl:173]: Only run on upgrade. For new install, `prometheus-operator-crd` need to be installed first.

**Comment by TheoBrigitte** [comment]
Patching works, but there's always more fun ...


```
$ k logs -f prometheus-operator-app-crd-install-shhs5 -nprometheus-operator-app
+ set -o nounset
+ exec
+ kubectl api-resources '--api-group=monitoring.coreos.com' -oname
+ tr '\n' ' '
+ xargs kubectl patch '--record=false' -p '{"metadata":{"annotations":{"meta.helm.sh/release-name": "prometheus-operator-crd","meta.helm.sh/release-namespace": "prometheus-operator-crd"},"labels":{"app.kubernetes.io/managed-by": "Helm"}}}' crd
Flag --record has been deprecated, --record will be removed in the future
I0118 16:31:52.420832      14 request.go:665] Waited for 1.18145944s due to client-side throttling, not priority and fairness, request: GET:https://172.31.0.1:443/apis/rbac.authorization.k8s.io/v1?timeout=32s
I0118 16:32:02.420978      14 request.go:665] Waited for 9.196336856s due to client-side throttling, not priority and fairness, request: GET:https://172.31.0.1:443/apis/crd.k8s.amazonaws.com/v1alpha1?timeout=32s
customresourcedefinition.apiextensions.k8s.io/alertmanagerconfigs.monitoring.coreos.com patched
customresourcedefinition.apiextensions.k8s.io/alertmanagers.monitoring.coreos.com patched
customresourcedefinition.apiextensions.k8s.io/podmonitors.monitoring.coreos.com patched
customresourcedefinition.apiextensions.k8s.io/probes.monitoring.coreos.com patched
customresourcedefinition.apiextensions.k8s.io/prometheuses.monitoring.coreos.com patched
customresourcedefinition.apiextensions.k8s.io/prometheusrules.monitoring.coreos.com patched
customresourcedefinition.apiextensions.k8s.io/servicemonitors.monitoring.coreos.com patched
customresourcedefinition.apiextensions.k8s.io/thanosrulers.monitoring.coreos.com patched
~ (gs-gauss-ac52g:default)$ k get crd alertmanagerconfigs.monitoring.coreos.com -oyaml|yq -y -r '.metadata|del(.managedFields,.annotations["kubectl.kubernetes.io/last-applied-configuration"])'|less
annotations:
  controller-gen.kubebuilder.io/version: v0.4.1
  meta.helm.sh/release-name: prometheus-operator-crd
  meta.helm.sh/release-namespace: prometheus-operator-crd
creationTimestamp: '2022-01-18T16:30:35Z'
generation: 1
labels:
  app.kubernetes.io/managed-by: Helm
name: alertmanagerconfigs.monitoring.coreos.com
resourceVersion: '17946'
uid: d9bd2013-0d9a-4d3e-b69b-9e6bf41b0cd0
```

**Comment by TheoBrigitte** [comment]
#### curent
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    meta.helm.sh/release-name: prometheus-operator-app
    meta.helm.sh/release-namespace: prometheus-operator-app
  labels:
    app.kubernetes.io/instance: prometheus-operator-app
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: kube-state-metrics
    app.kubernetes.io/version: 2.2.0
    helm.sh/chart: kube-state-metrics-3.5.2
  name: prometheus-operator-app-kube-state-metrics
  namespace: prometheus-operator-app
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/name: kube-state-metrics
```

#### desired
```yaml
# Source: prometheus-operator-app/charts/kube-state-metrics/templates/deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: prometheus-operator-app-kube-state-metrics
  namespace: prometheus-operator-app
  labels:
    helm.sh/chart: kube-state-metrics-4.1.2
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/component: metrics
    app.kubernetes.io/part-of: kube-state-metrics
    app.kubernetes.io/name: kube-state-metrics
    app.kubernetes.io/instance: prometheus-operator-app
    app.kubernetes.io/version: "2.2.4"
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: kube-state-metrics
      app.kubernetes.io/instance: prometheus-operator-app
```

**Comment by TheoBrigitte** [comment]
Ok everything works, both fixes apply and the release get deployed :heavy_check_mark: 

#### CRD fix
```
$ k logs -fn=prometheus-operator-app prometheus-operator-app-crd-install-hsqdv
+ set -o nounset
+ exec
+ kubectl api-resources '--api-group=monitoring.coreos.com' -oname
+ tr '\n' ' '
+ xargs kubectl patch '--record=false' -p '{"metadata":{"annotations":{"meta.helm.sh/release-name": "prometheus-operator-crd","meta.helm.sh/release-namespace": "prometheus-operator-crd"},"labels":{"app.kubernetes.io/managed-by": "Helm"}}}' crd
Flag --record has been deprecated, --record will be removed in the future
I0120 15:19:17.321747      15 request.go:665] Waited for 1.181657908s due to client-side throttling, not priority and fairness, request: GET:https://172.31.0.1:443/apis/storage.k8s.io/v1beta1?timeout=32s
I0120 15:19:27.322104      15 request.go:665] Waited for 9.197937913s due to client-side throttling, not priority and fairness, request: GET:https://172.31.0.1:443/apis/cert-manager.io/v1?timeout=32s
customresourcedefinition.apiextensions.k8s.io/alertmanagerconfigs.monitoring.coreos.com patched (no change)
customresourcedefinition.apiextensions.k8s.io/alertmanagers.monitoring.coreos.com patched (no change)
customresourcedefinition.apiextensions.k8s.io/podmonitors.monitoring.coreos.com patched (no change)
customresourcedefinition.apiextensions.k8s.io/probes.monitoring.coreos.com patched (no change)
customresourcedefinition.apiextensions.k8s.io/prometheuses.monitoring.coreos.com patched (no change)
customresourcedefinition.apiextensions.k8s.io/prometheusrules.monitoring.coreos.com patched (no change)
customresourcedefinition.apiextensions.k8s.io/servicemonitors.monitoring.coreos.com patched (no change)
customresourcedefinition.apiextensions.k8s.io/thanosrulers.monitoring.coreos.com patched (no change)
```

#### kube-state-metrics fix
```
$ k logs -fn=prometheus-operator-app prometheus-operator-app-ksm-fix-2hrbz
+ set -o nounset
+ exec
+ kubectl delete deployments.apps -l 'app.kubernetes.io/instance=prometheus-operator-app,app.kubernetes.io/name=kube-state-metrics' '--cascade=orphan'
I0120 15:19:33.728034       6 request.go:665] Waited for 1.184666127s due to client-side throttling, not priority and fairness, request: GET:https://172.31.0.1:443/apis/coordination.k8s.io/v1?timeout=32s
I0120 15:19:43.928485       6 request.go:665] Waited for 9.598223904s due to client-side throttling, not priority and fairness, request: GET:https://172.31.0.1:443/apis/crd.k8s.amazonaws.com/v1alpha1?timeout=32s
deployment.apps "prometheus-operator-app-kube-state-metrics" deleted
```

File: helm/prometheus-operator-app/values.yaml [hunk 2]
```diff
@@ -2709,6 +2709,3 @@ prometheus:
     ## https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#podmetricsendpoint
     ##
     # podMetricsEndpoints: []
-
-crds:
-  install: true
```
Comment (TheoBrigitte) [inline helm/prometheus-operator-app/values.yaml:2714]: This is a lie, crds are in a different chart now.

File: helm/prometheus-operator-app/templates/crd-install/kube-state-metrics-fix.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        - sh
+        - -c
+        - |
+          set -o errexit ; set -o xtrace ; set -o nounset
+
+          # piping stderr to stdout means kubectl's errors are surfaced
+          # in the pod's logs.
+          exec 2>&1
+
+          kubectl delete deployments.apps -l app.kubernetes.io/instance=prometheus-operator-app,app.kubernetes.io/name=kube-state-metrics --cascade=orphan
```
Comment (TheoBrigitte) [inline helm/prometheus-operator-app/templates/crd-install/kube-state-metrics-fix.yaml:41]: Found out that this is the official fix, see https://github.com/prometheus-community/helm-charts/blob/main/charts/prometheus/README.md#to-150

Comment (QuentinBisson) [reply helm/prometheus-operator-app/templates/crd-install/kube-state-metrics-fix.yaml:41]: Nice

**Comment by QuentinBisson** [comment]
Awesome Job Theo

**Review by QuentinBisson — APPROVED** [review]

File: helm/prometheus-operator-app/values.yaml (hunk 2 above)
Comment (QuentinBisson) [reply helm/prometheus-operator-app/values.yaml:2714]: The cake is a lie

Comment (TheoBrigitte) [reply helm/prometheus-operator-app/values.yaml:2714]: image

**Pull request merged**